core/src/main/java/io/github/lsposed/lspd/service/PackageService.java

```diff
@@ -176,4 +195,87 @@ private static PackageInfo getPackageInfoWithComponents(String packageName, int
+            @SuppressWarnings("JavaReflectionMemberAccess")
+            Constructor<IntentSender> intentSenderConstructor = IntentSender.class.getConstructor(IIntentSender.class);
+            intentSenderConstructor.setAccessible(true);
+            return intentSenderConstructor.newInstance(this);
+        }
+    }
+
+    private static void uninstallPackage(VersionedPackage versionedPackage) throws RemoteException, InterruptedException, InvocationTargetException, NoSuchMethodException, InstantiationException, IllegalAccessException {
+        CountDownLatch latch = new CountDownLatch(1);
+        pm.getPackageInstaller().uninstallExistingPackage(versionedPackage, "com.android.shell", new IntentSenderAdaptor() {
+            @Override
+            public void send(Intent intent) {
+                latch.countDown();
+            }
+        }.getIntentSender(), 0);
+        latch.await();
+    }
+
+    public static synchronized boolean installManagerIfAbsent(String packageName, File apkFile) {
+        IPackageManager pm = getPackageManager();
+        if (pm == null) return false;
+
+        try {
+            PackageInfo pkgInfo = pm.getPackageInfo(packageName, 0, 0);
+            if (pkgInfo != null && pkgInfo.versionName != null && pkgInfo.versionName.equals(BuildConfig.VERSION_NAME) && pkgInfo.applicationInfo != null && InstallerVerifier.verifyInstallerSignature(pkgInfo.applicationInfo))
+                return false;
+            // manager is not installed or version not matched, install stub
+            if (pkgInfo != null) {
+                uninstallPackage(new VersionedPackage(pkgInfo.packageName, pkgInfo.versionCode));
+            }
+            IPackageInstaller installerService = pm.getPackageInstaller();
+            String installerPackageName = "com.android.shell";
+            @SuppressWarnings("JavaReflectionMemberAccess")
+            Constructor<PackageInstaller> installerConstructor = PackageInstaller.class.getConstructor(IPackageInstaller.class, String.class, int.class);
+            installerConstructor.setAccessible(true);
+            PackageInstaller installer = installerConstructor.newInstance(installerService, installerPackageName, 0);
+            PackageInstaller.SessionParams params = new PackageInstaller.SessionParams(PackageInstaller.SessionParams.MODE_FULL_INSTALL);
+            int installFlags = HiddenApiBridge.PackageInstaller_SessionParams_installFlags(params);
+            installFlags |= 0x00000004/*PackageManager.INSTALL_ALLOW_TEST*/ | 0x00000002/*PackageManager.INSTALL_REPLACE_EXISTING*/;
+            HiddenApiBridge.PackageInstaller_SessionParams_installFlags(params, installFlags);
+
+            int sessionId = installer.createSession(params);
+            try (PackageInstaller.Session session = installer.openSession(sessionId)) {
+                try (InputStream is = new FileInputStream(apkFile); OutputStream os = session.openWrite(apkFile.getName(), 0, -1)) {
```

Comment:
seems that is and os is not closed after being used?

Reply:
https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html